dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/resolver/CategoryOptionGroupResolver.java

```diff
@@ -27,155 +27,119 @@
     }
 
-    private List<String> evaluate( String dataElementUid, String uid, String uid2 )
+    private String getExpression( String expression, DimensionalItemId id, List<String> cogUidList )
     {
-        List<String> resolvedExpression = new ArrayList<>();
-        Optional<String> cogUid = getCategoryOptionGroupUid( uid );
-        if ( cogUid.isPresent() )
+        List<String> cocUidIntersection = getCategoryOptionCombosIntersection( cogUidList );
+
+        if ( cocUidIntersection == null || cocUidIntersection.isEmpty() )
         {
-            Set<String> cocs = resolveCoCFromCog( cogUid.get() );
-            resolvedExpression = Arrays.asList( resolve( cocs, dataElementUid, uid2 ).split( "\\+" ) );
+            return expression;
         }
-        return resolvedExpression;
-    }
 
-    private String resolve( Set<String> cocs, String dataElementUid, String third )
-    {
-        boolean isAoc = isAoc( third );
+        List<String> resolved = cocUidIntersection
+            .stream()
+            .map( cocUid -> id.getItem().replace( id.getId1(), cocUid ) )
+            .collect( Collectors.toList() );
 
-        return cocs.stream()
-            .map( coc -> EXP_OPEN + dataElementUid + SEPARATOR + coc + (isAoc ? SEPARATOR + third : "") + EXP_CLOSE )
-            .collect( Collectors.joining( "+" ) );
-    }
+        expression = expression.replace( id.getItem(),
+            LEFT_BRACKET + Joiner.on( "+" ).join( resolved ) + RIGHT_BRACKET );
 
-    private boolean isAoc( String uid )
-    {
-        return (uid != null && categoryOptionComboStore.getByUid( uid ) != null);
+        return expression;
     }
 
-    private Optional<String> getCategoryOptionGroupUid( String uid )
+    private List<String> getCategoryOptionCombosIntersection( List<String> cogUidList )
     {
-        CategoryOptionGroup cog = categoryOptionGroupStore.getByUid( uid );
+        List<String> cocUidIntersection = null;
 
-        return cog == null ? Optional.empty() : Optional.of( cog.getUid() );
-    }
+        for ( String cogUid : cogUidList )
+        {
+            CategoryOptionGroup cog = categoryOptionGroupStore
+                .getByUid( cogUid );
+
+            if ( cog != null )
+            {
+                if ( cocUidIntersection == null )
+                {
+                    cocUidIntersection = categoryOptionComboStore.getCategoryOptionCombosByGroupUid( cog.getUid() )
+                        .stream()
+                        .map( BaseIdentifiableObject::getUid )
+                        .collect( Collectors.toList() );
+                }
+                else
+                {
+                    cocUidIntersection
+                        .retainAll( categoryOptionComboStore.getCategoryOptionCombosByGroupUid( cog.getUid() )
+                            .stream()
+                            .map( BaseIdentifiableObject::getUid )
+                            .collect( Collectors.toList() ) );
+                }
```

Comment:
You could reduce redundancy (and not make the reader scan to see if the statements are similar) with:

                List<String> cocUids = categoryOptionComboStore.getCategoryOptionCombosByGroupUid( cog.getUid() )
                    .stream()
                    .map( BaseIdentifiableObject::getUid )
                    .collect( Collectors.toList() );

                if ( cocUidIntersection == null )
                {
                    cocUidIntersection = cocUids;
                }
                else
                {
                    cocUidIntersection.retainAll( cocUids );
                }

Reply:
Done